File: testsuite/simulation/deleteConnector1.lua
-- status: correct
-- teardown_command: rm -rf deleteConnector1_lua/
-- linux: yes
-- mingw32: yes
-- mingw64: yes
-- win: no
-- mac: yes

oms_setCommandLineOption("--suppressPath=true")
oms_setTempDirectory("./deleteConnector1_lua/")
oms_setWorkingDirectory("./deleteConnector1_lua/")

oms_newModel("deleteConnector")

oms_addSystem("deleteConnector.Root", oms_system_wc)

oms_addResources("deleteConnector.Root:deleteConnector.ssv")

oms_addConnector("deleteConnector.Root.C1", oms_causality_input, oms_signal_type_real)
oms_addConnector("deleteConnector.Root.C2", oms_causality_output, oms_signal_type_real)
oms_setReal("deleteConnector.Root.C1", -70.0)

oms_addConnection("deleteConnector.Root.C1", "deleteConnector.Root.C2")

oms_addSystem("deleteConnector.Root.System1", oms_system_sc)
oms_addConnector("deleteConnector.Root.System1.C1", oms_causality_input, oms_signal_type_real)
oms_addConnector("deleteConnector.Root.System1.C2", oms_causality_input, oms_signal_type_real)

oms_addSystem("deleteConnector.Root.System2", oms_system_sc)
oms_addConnector("deleteConnector.Root.System2.C3", oms_causality_output, oms_signal_type_real)
oms_addConnector("deleteConnector.Root.System2.C4", oms_causality_output, oms_signal_type_real)
oms_setReal("deleteConnector.Root.System2.C3", 20.0)
oms_setReal("deleteConnector.Root.System2.C4", 30.0)

oms_addSubModel("deleteConnector.Root.System1.Gain", "../../resources/Modelica.Blocks.Math.Gain.fmu")
oms_setReal("deleteConnector.Root.System1.Gain.k", 30.0)

oms_addConnection("deleteConnector.Root.System2.C3", "deleteConnector.Root.System1.C1")
oms_addConnection("deleteConnector.Root.System2.C4", "deleteConnector.Root.System1.C2")

oms_setResultFile("deleteConnector", "deleteConnector1.mat", 10)

src = oms_exportSnapshot("deleteConnector")
print(src)

oms_delete("deleteConnector.Root.C1")

-- delete connector "deleteConnector.Root.System2.C3" and associated connection
oms_delete("deleteConnector.Root.System2.C3")

-- delete components connector and its startValue
oms_delete("deleteConnector.Root.System1.Gain.k")

-- delete unknown connector signal for warning
oms_delete("deleteConnector.Root.System2.C6")

-- delete unknown system for warning
oms_delete("deleteConnector.Root.System3")

-- after deletion
src = oms_exportSnapshot("deleteConnector:resources/deleteConnector.ssv")
print(src)

oms_terminate("deleteConnector")
oms_delete("deleteConnector")


-- Result:
-- <?xml version="1.0"?>
-- <oms:snapshot
--   xmlns:oms="https://raw.githubusercontent.com/OpenModelica/OMSimulator/master/schema/oms.xsd"
--   partial="false">
--   <oms:file
--     name="SystemStructure.ssd">
--     <ssd:SystemStructureDescription
--       xmlns:ssc="http://ssp-standard.org/SSP1/SystemStructureCommon"
--       xmlns:ssd="http://ssp-standard.org/SSP1/SystemStructureDescription"
--       xmlns:ssv="http://ssp-standard.org/SSP1/SystemStructureParameterValues"
--       xmlns:ssm="http://ssp-standard.org/SSP1/SystemStructureParameterMapping"
--       xmlns:ssb="http://ssp-standard.org/SSP1/SystemStructureSignalDictionary"
--       xmlns:oms="https://raw.githubusercontent.com/OpenModelica/OMSimulator/master/schema/oms.xsd"
--       name="deleteConnector"
--       version="1.0">
--       <ssd:System
--         name="Root">
--         <ssd:Connectors>
--           <ssd:Connector
--             name="C1"
--             kind="input">
--             <ssc:Real />
--           </ssd:Connector>
--           <ssd:Connector
--             name="C2"
--             kind="output">
--             <ssc:Real />
--           </ssd:Connector>
--         </ssd:Connectors>
--         <ssd:ParameterBindings>
--           <ssd:ParameterBinding
--             source="resources/deleteConnector.ssv" />
--         </ssd:ParameterBindings>
--         <ssd:Elements>
--           <ssd:System
--             name="System2">
--             <ssd:Connectors>
--               <ssd:Connector
--                 name="C3"
--                 kind="output">
--                 <ssc:Real />
--               </ssd:Connector>
--               <ssd:Connector
--                 name="C4"
--                 kind="output">
--                 <ssc:Real />
--               </ssd:Connector>
--             </ssd:Connectors>
--             <ssd:Annotations>
--               <ssc:Annotation
--                 type="org.openmodelica">
--                 <oms:Annotations>
--                   <oms:SimulationInformation>
--                     <oms:VariableStepSolver
--                       description="cvode"
--                       absoluteTolerance="0.0001"
--                       relativeTolerance="0.0001"
--                       minimumStepSize="1e-12"
--                       maximumStepSize="0.001"
--                       initialStepSize="1e-06" />
--                   </oms:SimulationInformation>
--                 </oms:Annotations>
--               </ssc:Annotation>
--             </ssd:Annotations>
--           </ssd:System>
--           <ssd:System
--             name="System1">
--             <ssd:Connectors>
--               <ssd:Connector
--                 name="C1"
--                 kind="input">
--                 <ssc:Real />
--               </ssd:Connector>
--               <ssd:Connector
--                 name="C2"
--                 kind="input">
--                 <ssc:Real />
--               </ssd:Connector>
--             </ssd:Connectors>
--             <ssd:Elements>
--               <ssd:Component
--                 name="Gain"
--                 type="application/x-fmu-sharedlibrary"
--                 source="resources/0001_Gain.fmu">
--                 <ssd:Connectors>
--                   <ssd:Connector
--                     name="u"
--                     kind="input">
--                     <ssc:Real />
--                     <ssd:ConnectorGeometry
--                       x="0.000000"
--                       y="0.500000" />
--                   </ssd:Connector>
--                   <ssd:Connector
--                     name="y"
--                     kind="output">
--                     <ssc:Real />
--                     <ssd:ConnectorGeometry
--                       x="1.000000"
--                       y="0.500000" />
--                   </ssd:Connector>
--                   <ssd:Connector
--                     name="k"
--                     kind="parameter">
--                     <ssc:Real />
--                   </ssd:Connector>
--                 </ssd:Connectors>
--               </ssd:Component>
--             </ssd:Elements>
--             <ssd:Annotations>
--               <ssc:Annotation
--                 type="org.openmodelica">
--                 <oms:Annotations>
--                   <oms:SimulationInformation>
--                     <oms:VariableStepSolver
--                       description="cvode"
--                       absoluteTolerance="0.0001"
--                       relativeTolerance="0.0001"
--                       minimumStepSize="1e-12"
--                       maximumStepSize="0.001"
--                       initialStepSize="1e-06" />
--                   </oms:SimulationInformation>
--                 </oms:Annotations>
--               </ssc:Annotation>
--             </ssd:Annotations>
--           </ssd:System>
--         </ssd:Elements>
--         <ssd:Connections>
--           <ssd:Connection
--             startElement=""
--             startConnector="C1"
--             endElement=""
--             endConnector="C2" />
--           <ssd:Connection
--             startElement="System2"
--             startConnector="C3"
--             endElement="System1"
--             endConnector="C1" />
--           <ssd:Connection
--             startElement="System2"
--             startConnector="C4"
--             endElement="System1"
--             endConnector="C2" />
--         </ssd:Connections>
--         <ssd:Annotations>
--           <ssc:Annotation
--             type="org.openmodelica">
--             <oms:Annotations>
--               <oms:SimulationInformation>
--                 <oms:FixedStepMaster
--                   description="oms-ma"
--                   stepSize="0.001000"
--                   absoluteTolerance="0.000100"
--                   relativeTolerance="0.000100" />
--               </oms:SimulationInformation>
--             </oms:Annotations>
--           </ssc:Annotation>
--         </ssd:Annotations>
--       </ssd:System>
--       <ssd:DefaultExperiment
--         startTime="0.000000"
--         stopTime="1.000000">
--         <ssd:Annotations>
--           <ssc:Annotation
--             type="org.openmodelica">
--             <oms:Annotations>
--               <oms:SimulationInformation
--                 resultFile="deleteConnector1.mat"
--                 loggingInterval="0.000000"
--                 bufferSize="10"
--                 signalFilter="resources/signalFilter.xml" />
--             </oms:Annotations>
--           </ssc:Annotation>
--         </ssd:Annotations>
--       </ssd:DefaultExperiment>
--     </ssd:SystemStructureDescription>
--   </oms:file>
--   <oms:file
--     name="resources/deleteConnector.ssv">
--     <ssv:ParameterSet
--       xmlns:ssc="http://ssp-standard.org/SSP1/SystemStructureCommon"
--       xmlns:ssv="http://ssp-standard.org/SSP1/SystemStructureParameterValues"
--       version="1.0"
--       name="parameters">
--       <ssv:Parameters>
--         <ssv:Parameter
--           name="System2.C4">
--           <ssv:Real
--             value="30" />
--         </ssv:Parameter>
--         <ssv:Parameter
--           name="System2.C3">
--           <ssv:Real
--             value="20" />
--         </ssv:Parameter>
--         <ssv:Parameter
--           name="Gain.k">
--           <ssv:Real
--             value="30" />
--         </ssv:Parameter>
--         <ssv:Parameter
--           name="C1">
--           <ssv:Real
--             value="-70" />
--         </ssv:Parameter>
--       </ssv:Parameters>
--     </ssv:ParameterSet>
--   </oms:file>
--   <oms:file
--     name="resources/signalFilter.xml">
--     <oms:SignalFilter
--       version="1.0">
--       <oms:Variable
--         name="deleteConnector.Root.C1"
--         type="Real"
--         kind="input" />
--       <oms:Variable
--         name="deleteConnector.Root.C2"
--         type="Real"
--         kind="output" />
--       <oms:Variable
--         name="deleteConnector.Root.System2.C3"
--         type="Real"
--         kind="output" />
--       <oms:Variable
--         name="deleteConnector.Root.System2.C4"
--         type="Real"
--         kind="output" />
--       <oms:Variable
--         name="deleteConnector.Root.System1.C1"
--         type="Real"
--         kind="input" />
--       <oms:Variable
--         name="deleteConnector.Root.System1.C2"
--         type="Real"
--         kind="input" />
--       <oms:Variable
--         name="deleteConnector.Root.System1.Gain.u"
--         type="Real"
--         kind="input" />
--       <oms:Variable
--         name="deleteConnector.Root.System1.Gain.y"
--         type="Real"
--         kind="output" />
--       <oms:Variable
--         name="deleteConnector.Root.System1.Gain.k"
--         type="Real"
--         kind="parameter" />
--     </oms:SignalFilter>
--   </oms:file>
-- </oms:snapshot>
--
-- warning: failed to delete object "deleteConnector.Root.System2.C6" because the identifier couldn't be resolved to any connector, component, system, or model
-- warning: failed to delete object "deleteConnector.Root.System3" because the identifier couldn't be resolved to any connector, component, system, or model
-- <?xml version="1.0"?>
-- <oms:snapshot
--   xmlns:oms="https://raw.githubusercontent.com/OpenModelica/OMSimulator/master/schema/oms.xsd"
--   partial="true">
--   <oms:file
--     name="resources/deleteConnector.ssv">
--     <ssv:ParameterSet
--       xmlns:ssc="http://ssp-standard.org/SSP1/SystemStructureCommon"
--       xmlns:ssv="http://ssp-standard.org/SSP1/SystemStructureParameterValues"
--       version="1.0"
--       name="parameters">
--       <ssv:Parameters>
--         <ssv:Parameter
--           name="System2.C4">
--           <ssv:Real
--             value="30" />
--         </ssv:Parameter>
--       </ssv:Parameters>
--     </ssv:ParameterSet>
--   </oms:file>
-- </oms:snapshot>
--
-- info:    2 warnings
-- info:    0 errors
-- endResult

File: testsuite/api/deleteConnector1.py
## status: correct
## teardown_command: rm deleteConnector1.ssp
## linux: yes
## ucrt64: yes
## win: yes
## mac: yes

from OMSimulator import SSP, CRef, Settings, Connector, Causality, SignalType

Settings.suppressPath = True

model = SSP()
model.addResource('../resources/Modelica.Blocks.Math.Add.fmu', new_name='resources/Add.fmu')
model.addResource('../resources/Modelica.Blocks.Math.Gain.fmu', new_name='resources/Gain.fmu')

model.addSystem(CRef('default', 'sub-system'))
## add top level system connector
model.activeVariant.system.addConnector(Connector('param1', Causality.parameter, SignalType.Real))
## add top level sub-system connector
model.activeVariant.system.elements[CRef('sub-system')].addConnector(Connector('input', Causality.input, SignalType.Real))

component1 = model.addComponent(CRef('default', 'Add1'), 'resources/Add.fmu')
component2 = model.addComponent(CRef('default', 'sub-system', 'Add2'), 'resources/Add.fmu')

solver1 = {'name' : 'solver1',  'method': 'euler', 'tolerance': 1e-6}
model.newSolver(solver1)

model.setSolver(CRef('default', 'Add1'), 'solver1')
model.setSolver(CRef('default', 'sub-system', 'Add2'), 'solver1')

model.addConnection(CRef('default', 'param1'), CRef('default', 'Add1', 'u1'))
model.addConnection(CRef('default', 'Add1', 'y'), CRef('default', 'sub-system', 'input'))
model.addConnection(CRef('default', 'sub-system', 'Add2', 'u1'), CRef('default', 'sub-system', 'input'))

model.export('deleteConnector1.ssp')

model2 = SSP('deleteConnector1.ssp')
model2.delete(CRef('default', 'param1'))
model2.delete(CRef('default', 'sub-system', 'input'))
model2.delete(CRef('default', 'sub-system', 'Add2', 'u1'))
model2.delete(CRef('default', 'Add1', 'u1'))
print("## After deleting connectors:")
print("=============================")
model2.list()

## Result:
## ## After deleting connectors:
## =============================
## <class 'OMSimulator.ssp.SSP'>
## |-- Resources:
## |--   resources/Add.fmu
## |--   resources/Gain.fmu
## |-- Active Variant: default
## |-- <class 'OMSimulator.ssd.SSD'>
## |-- Variant "default": <hidden>
## |-- |-- System: default 'None'
## |-- |-- |-- Connectors:
## |-- |-- |-- Elements:
## |-- |-- |-- |-- System: sub-system 'None'
## |-- |-- |-- |-- |-- Connectors:
## |-- |-- |-- |-- |-- Elements:
## |-- |-- |-- |-- |-- |-- FMU: Add2 'None'
## |-- |-- |-- |-- |-- |-- |-- path: resources/Add.fmu
## |-- |-- |-- |-- |-- |-- |-- Connectors:
## |-- |-- |-- |-- |-- |-- |-- |-- (u2, Causality.input, SignalType.Real, None, 'Connector of Real input signal 2')
## |-- |-- |-- |-- |-- |-- |-- |-- (y, Causality.output, SignalType.Real, None, 'Connector of Real output signal')
## |-- |-- |-- |-- |-- |-- |-- |-- (k1, Causality.parameter, SignalType.Real, None, 'Gain of input signal 1')
## |-- |-- |-- |-- |-- |-- |-- |-- (k2, Causality.parameter, SignalType.Real, None, 'Gain of input signal 2')
## |-- |-- |-- |-- |-- |-- |-- Solver Settings:
## |-- |-- |-- |-- |-- |-- |-- |-- name: solver1
## |-- |-- |-- |-- |-- Connections:
## |-- |-- |-- |-- |-- |-- Add2.u1 -> .input
## |-- |-- |-- |-- FMU: Add1 'None'
## |-- |-- |-- |-- |-- path: resources/Add.fmu
## |-- |-- |-- |-- |-- Connectors:
## |-- |-- |-- |-- |-- |-- (u2, Causality.input, SignalType.Real, None, 'Connector of Real input signal 2')
## |-- |-- |-- |-- |-- |-- (y, Causality.output, SignalType.Real, None, 'Connector of Real output signal')
## |-- |-- |-- |-- |-- |-- (k1, Causality.parameter, SignalType.Real, None, 'Gain of input signal 1')
## |-- |-- |-- |-- |-- |-- (k2, Causality.parameter, SignalType.Real, None, 'Gain of input signal 2')
## |-- |-- |-- |-- |-- Solver Settings:
## |-- |-- |-- |-- |-- |-- name: solver1
## |-- |-- |-- Connections:
## |-- |-- |-- |-- .param1 -> Add1.u1
## |-- |-- |-- |-- Add1.y -> sub-system.input
## |-- |-- |-- Solver Settings:
## |-- |-- |-- |-- (name=solver1, method=euler, tolerance=1e-06)
## |-- DefaultExperiment
## |-- |-- startTime: 0.0
## |-- |-- stopTime: 1.0
## endResult
